main.js

```diff
@@ -0,0 +1,242 @@
+class WeatherApp {
+  constructor() {
+    this.coordinates = {};
+    this.weatherInfos = {};
+    this._init();
+  }
+
+  _init() {
+    const searchForm = document.querySelector('.weather__search-form');
+    searchForm.addEventListener('submit', event => {
+      const locationInput = document.querySelector('.weather__textbox').value;
+      event.preventDefault();
+      if (!locationInput) return;
+      this._getCoordinates(locationInput);
+    });
+  }
+
+  async _getCoordinates(location) {
+    try {
+      this._displayLoader();
+      //TODO：Fetchを別のメソッドに切り出す
+      const response = await fetch(`http://api.openweathermap.org/geo/1.0/direct?q=${location}&limit=10&appid=${APIKEY}`);
+
+      if (!response.ok) {
+        this._errorThower(response.status);
+      }
+
+      const locationDetailsArray = await response.json();
+
+      if (locationDetailsArray.length === 0) {
+        throw new ErrorNoCity('都市が見つかりませんでした。')
+      }
+
+      if (locationDetailsArray.length >= 2) {
+        this._getAllLocationNames(locationDetailsArray);
+        return;
+      }
+
+      this.coordinates = {
+        lat: locationDetailsArray[0].lat,
+        lon: locationDetailsArray[0].lon,
+      }
+
+      await this._getWeather(this.coordinates);
+
+    } catch (e) {
+      this._resetHtml();
+      new ErrorHandler(e);
+
+    } finally {
+      this._hideLoader();
+    }
+
+  }
+
+  _getAllLocationNames(coordArr) {
+    const allLocations = coordArr.map(location => this._extractLocationInfo(location));
+    const uniqueLocations = this._discardIdenticalNamedLocations(allLocations)
+
+    if (uniqueLocations.length === 1) {
+      this._getWeather(uniqueLocations);
+
+    } else {
+      this._renderLocationSuggestions(uniqueLocations);
+    }
+  }
+
+  //都市の日本語名が表示可能な場合は日本語名の表示を優先し、なければ英語名を表示する
+  _extractLocationInfo(location) {
+    const locationInfo = {
+      name: location.name,
+      lat: location.lat,
+      lon: location.lon,
+    }
+    if (location.local_names && location.local_names.ja) {
+      locationInfo.name = location.local_names.ja;
+    }
+    return locationInfo;
+  }
+
+  //重複した都市名を排除する
+  _discardIdenticalNamedLocations(multiLoc) {
+    const uniqueLocationsMap = Array.from(new Map(multiLoc.map(location => [location.name, location])));
+    return uniqueLocationsMap.map(location => location[1]);
+  }
```

Comment:
```suggestion
  //都市名が重複してAPIから返却される場合があるので、事前に重複を排除する
  _removeDuplicateLocations(locations) {
    const uniqueLocationsMap = Array.from(new Map(multiLoc.map(location => [location.name, location])));
    return uniqueLocationsMap.map(location => location[1]);
  }
```
僕ならこうするを書いておきます
- コメントもうちょい踏み込んでwhyをかく
- メソッド名はシンプルにできるはず。難しいメソッドは難しいことしているべきで、これって非常に単純なことしてるだけだよね？

Reply:
[MUST]
```
const uniqueLocationsMap = Array.from(new Map(multiLoc.map(location => [location.name, location])));
```
この結果はMapではないので嘘ついてます。確認して実装してください。

Reply:
```
uniqueLocationsMap.map(location => location[1]);
```
重複を排除する以外のこともやってませんかね？

Reply:
@cyndiii29 
https://github.com/KenJamesHioki/weather_app/commit/d3e7bfd8a807b40e37435288e73c8ec625cd43fd

[変更点]
・コメント内容の変更
・メソッド名の変更
・uniqueLocationsMapの命名変更

[IMO]
> ```
> uniqueLocationsMap.map(location => location[1]);
> ```
> 
> 重複を排除する以外のこともやってませんかね？
メソッドの引数と同じ形式でreturnをしているので問題ないと考えています。

一個前のコードで
`const uniqueLocationsMap = Array.from(new Map(multiLoc.map(location => [location.name, location])));
`
形式を変えてしまっているので、returnする前に形式を戻してあげています。